File: Data Structures & Algorithms/remove-node-from-end-of-linked-list/submission-1.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* removeNthFromEnd(ListNode* head, int n) {
        if(!head) return head;
        if(head->next==nullptr && n==1) return nullptr;
        ListNode* temp = head;
        int i = 0;
        while(temp!=nullptr){
            temp = temp->next;
            i++;

        }
        temp = head;
        if(i==n){
            head = head->next;
            return head;
        }
        while(i!=n+1){
            temp=temp->next;
            i--;
        }
        temp->next = temp->next->next;
        
        return head;
    }
};

File: Data Structures & Algorithms/merge-k-sorted-linked-lists/submission-1.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:

    def Merge(self,list1 : ListNode,list2 : ListNode):
        dummy = ListNode(0,None)
        curr = dummy

        while list1 and list2:
            if list1.val >= list2.val:
                curr.next = list2
                list2 = list2.next
            else:
                curr.next = list1
                list1 = list1.next
            
            curr = curr.next

        curr.next = list1 if list1 else list2

        return dummy.next
    def mergeKLists(self, lists: List[Optional[ListNode]]) -> Optional[ListNode]:
        
        while len(lists)>1:
            merged = []
            for i in range(0,len(lists),2):
                list1 = lists[i]
                list2 = lists[i+1] if i+1 < len(lists) else None
                merged.append(self.Merge(list1,list2))

            lists = merged

        return lists[0] if lists else None 





        